### PR TITLE
feat: add get study config endpoint

### PR DESCRIPTION
**Pull request by Spazzy757**
# Changelog

- adds get study config endpoint

Fixes: https://github.com/vlab-research/vlab/issues/124

### CODE REVIEW

File: dashboard-api/internal/platform/server/handler/studyconf/read.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+			return
+		}
+
+		s, err := r.Study.GetStudyBySlug(ctx, req.Slug, auth.GetUserIdFrom(ctx))
+		if err != nil {
+			handleError(ctx, err)
+			return
+		}
+
+		dscs, err := r.StudyConf.GetByStudyID(ctx, s.Id)
```
Comment (nandanrao) [inline dashboard-api/internal/platform/server/handler/studyconf/read.go:36]: Why not just create a GetByStudySlug (and user)? Seems more efficient to just put that join in the query, no?

File: dashboard-api/internal/types/studyconf.go [hunk 2]
```diff
@@ -46,28 +49,50 @@ func (sc *StudyConf) TransformForDatabase() (res []DatabaseStudyConf, err error)
...
 }
 
+// TransformFromDatabase gets a list of DatabaseStudyConf and iterates
+// through it, adding the relevant configs to the StudyConf struct
+func (sc *StudyConf) TransformFromDatabase(dcs []*DatabaseStudyConf) error {
+	for _, d := range dcs {
+		//TODO: Im sure we can make this more future proof by using
+		//reflect and checking the json tags
+		switch d.ConfType {
+		case "general":
```
Comment (nandanrao) [inline dashboard-api/internal/types/studyconf.go:79]: I would make a map just to make it slightly cleaner, as we'll keep adding, and the logic will be identical.

**Comment by nandanrao** [comment]
Left some comments, nothing blocking. Overall, I approve!

File: dashboard-api/internal/platform/server/handler/studyconf/read_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+func TestHandler_StudyConfiguration_GetByStudySlug(t *testing.T) {
+	assert := require.New(t)
+	id := uuid.New()
+	studyslug := id.String()
+	testcases := []struct {
+		databasestudyconfs []*types.DatabaseStudyConf
+		studyslug          string
+		description        string
```
Comment (sssoja) [inline dashboard-api/internal/platform/server/handler/studyconf/read_test.go:22]: Not a feedback just a comment that this is the struct I was referring to earlier when I was redundantly passing the `slug` and the `description` from the frontend!

My understanding now is that the client is simply sending the json data that exists inside the `databasestudyconfs` and that this struct is just a concept for storing response data in the backend

File: dashboard-api/internal/testhelpers/testhelpers.go [hunk 4]
```diff
@@ -136,12 +137,20 @@ func DeleteAllUsers() {
...
 }
 
 func CreateUser() {
 	r := GetRepositories()
 	_, _ = r.Db.Exec("INSERT INTO users (id) VALUES ($1)", CurrentUserId)
 }
+
+func CreateDatabaseStudyConf(dsc types.DatabaseStudyConf) error {
+
+	r := GetRepositories()
```
Comment (sssoja) [inline dashboard-api/internal/testhelpers/testhelpers.go:152]: excuse my silly Go question: is that `r` for respositories and `q` for query?

File: dashboard-api/internal/platform/server/handler/studyconf/read.go (hunk 1 above)
Comment (Spazzy757) [reply dashboard-api/internal/platform/server/handler/studyconf/read.go:36]: True, I can update that

File: dashboard-api/internal/platform/server/handler/studyconf/read_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	"github.com/vlab-research/vlab/dashboard-api/internal/platform/storage"
+	"github.com/vlab-research/vlab/dashboard-api/internal/testhelpers"
+	"github.com/vlab-research/vlab/dashboard-api/internal/types"
+)
+
+func TestHandler_StudyConfiguration_GetByStudySlug(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Spazzy757) [reply dashboard-api/internal/platform/server/handler/studyconf/read_test.go:15]: Maybe we can discuss this in stand up, im not sure I understand what you mean :sweat_smile: There is only one test for the Get endpoint which does end-to-end

File: dashboard-api/internal/platform/server/handler/studyconf/read_test.go (hunk 3 above)
Comment (Spazzy757) [reply dashboard-api/internal/platform/server/handler/studyconf/read_test.go:22]: Ah right, no this structure is specifically what the database expects and not what the forntend would pass, its the backends job to convert what the frontend passes into this

File: dashboard-api/internal/testhelpers/testhelpers.go (hunk 4 above)
Comment (Spazzy757) [reply dashboard-api/internal/testhelpers/testhelpers.go:152]: yes, this is definitely not one of my favorite practices in Golang, but they tend to push towards very simple variables names that understanding can be pulled from the context of the decleration, in this case:
```golang
// We can assume that r is now a repositories structure due to the naming of the function
r  :=  GetRepositories()
```
There is quite a large [styleguide](https://google.github.io/styleguide/go/best-practices#variable-declarations) for golang written by google:

File: dashboard-api/internal/types/studyconf.go [hunk 6]
```diff
@@ -46,28 +49,50 @@ func (sc *StudyConf) TransformForDatabase() (res []DatabaseStudyConf, err error)
 		// service use to identify the config
 		fieldName := v.Type().Field(i).Tag.Get("json")
 		// if a field is not set we dont need to add it to our array
-		// as it will just have empty data
-		if field.IsZero() {
+		// as it will just have empty data as well as if it is not one of
+		// the confs which should all be pointers
+		if field.IsZero() || field.Kind() != reflect.Ptr {
```
[the remark this answers is not in the document]
Comment (Spazzy757) [reply dashboard-api/internal/types/studyconf.go:54]: so `Kind()` is essentially the type of the structure so in this case we are checking if its a type `Pointer` (you can also have types like `string` or `int` etc)

File: dashboard-api/internal/platform/storage/studyconf_repository.go [hunk 7]
```diff
@@ -43,3 +43,37 @@ func (r *StudyConfRepository) Create(
 
 	return nil
 }
+
+// GetConfsByStudy will return all the latest unique confs that are attached to
+// a study
+func (r *StudyConfRepository) GetByStudyID(
```
[the remark this answers is not in the document]
Comment (Spazzy757) [reply dashboard-api/internal/platform/storage/studyconf_repository.go:49]: I think this is what Nandan is mentioning earlier. In the database slug is not unique so but it makes sens that it should be unqiue

**Pull request merged**